File: back/node/app/core/service/game/GameLogic.ts
import {Result, success, error, notFound} from "../../common/result/Result.js";
import { PlayerType } from "../../model/PlayerType.js";
import {BoardType} from "../../model/BoardType.js";
import {StateType} from "../../model/StateType.js";
import {randomUUID} from "crypto";
import { CellType } from "../../model/CellType.js";
import {GameResult} from "../../model/GameResult.js";
import {GameStateCheckerB} from "./GameStateChecker.js";

interface GameLogicB {
    generateNewID(): Result<string>;
    generateNewBoard(): BoardType;
    randomPlayer(): Result<PlayerType>;
    getNextPlayer(currentPlayer: PlayerType): Result<PlayerType>;
    checkForWinner(boardState: StateType): Result<GameResult>;
}

class GameLogic implements GameLogicB {

    private checker: GameStateCheckerB

    constructor(gameStateChecker: GameStateCheckerB) {
        this.checker = gameStateChecker
    }

    checkForWinner(boardState: StateType): Result<GameResult> {

        if ('cells' in boardState) {

            const cells = boardState.cells;

            if(Array.isArray(cells)){
                return this.checker.checkForVictoryOrDrawA(cells)
            } else {
                return this.checker.checkForVictoryOrDrawB(cells)
            }

        } else {
            return this.checker.checkForVictoryOrDrawC(boardState)
        }

    }

    generateNewBoard(): BoardType {
        return {
                cells: [
                    [null, null, null],
                    [null, null, null],
                    [null, null, null]
                ]
            };
    }

    generateNewID(): Result<string> {
        const newID: string = randomUUID()
        return newID ? success(newID) : error("could not generate a new ID")
    }

    getNextPlayer(currentPlayer: PlayerType): Result<PlayerType> {
        const nextPlayer: PlayerType = currentPlayer.symbol === 'X' ? {symbol: 'O'} : {symbol: 'X'}
        return nextPlayer ? success(nextPlayer) : error("could not get player")
    }

    randomPlayer(): Result<PlayerType> {
        const players: NonNullable<CellType>[] = ['X', 'O']
        const [symbol ] = players.sort(() => Math.random() - 0.5);
        return symbol ? success({symbol}) : error("could not get player");
    }

}

export {GameLogic}

File: back/kotlin/app/core/service/game/GameLogic.kt
package core.service.game

import java.util.UUID

interface GameLogicB {
    fun generateNewID(): Result<String>
    fun generateNewBoard(): Result<BoardType>
    fun randomPlayer(): Result<PlayerType>
    fun getNextPlayer(currentPlayer: PlayerType): Result<PlayerType>
    fun checkForWinner(state: StateType): Result<GameResult>
}

class GameLogic(private val checker: GameStateCheckerB): GameLogicB {

    override fun generateNewID(): Result<String> {
        val newID = UUID.randomUUID().toString()
        return Result.Success(newID)
    }

    override fun generateNewBoard(): Result<BoardType> {
        return Result.Success(BoardType.ArrayBoard(Array(3) { Array(3) { null } }))
    }

    override fun randomPlayer(): Result<PlayerType> {
        val symbols = listOf(CellType.X, CellType.O).shuffled()
        val symbol = symbols.firstOrNull() ?: return Result.Error(IllegalStateException("Could not get symbol"))
        return Result.Success(PlayerType(symbol))
    }

    override fun getNextPlayer(currentPlayer: PlayerType): Result<PlayerType> {
        val nextSymbol = if (currentPlayer.symbol == CellType.X) CellType.O else CellType.X
        return Result.Success(PlayerType(nextSymbol))
    }

    override fun checkForWinner(state: StateType): Result<GameResult> {

        return when (state) {
            is State.BoardState -> {
                when (val boardType = state.board) {
                    is BoardType.ArrayBoard -> checker.checkForVictoryOrDrawA(boardType.cells)
                    is BoardType.DictionaryBoard -> checker.checkForVictoryOrDrawB(boardType.cells)
                }
            }
            is State.MovesState -> {
                checker.checkForVictoryOrDrawC(state.moves)
            }
        }

    }

}